streaming/protocol/setup.c

```diff
@@ -0,0 +1,283 @@
+        return -3;
+    else if (!memcmp(http, START_STREAMING_ERROR_NOT_PERMITTED, sizeof(START_STREAMING_ERROR_NOT_PERMITTED)))
+        return -4;
+    else
+        return -1;
+}
+
+static long int parse_stream_version(RRDHOST *host, const char *http)
+{
+    long int stream_version = -1;
+    int answer = -1;
+    char *stream_version_start = strchr(http, '=');
+    if (stream_version_start) {
+        stream_version_start++;
+        stream_version = strtol(stream_version_start, NULL, 10);
+        answer = memcmp(http, START_STREAMING_PROMPT_VN, (size_t)(stream_version_start - http));
+        if (!answer) {
+            rrdpush_set_flags_to_newest_stream(host);
+        }
+    } else {
+        answer = memcmp(http, START_STREAMING_PROMPT_V2, strlen(START_STREAMING_PROMPT_V2));
+        if (!answer) {
+            stream_version = 1;
+            rrdpush_set_flags_to_newest_stream(host);
+        } else {
+            answer = memcmp(http, START_STREAMING_PROMPT, strlen(START_STREAMING_PROMPT));
+            if (!answer) {
+                stream_version = 0;
+                rrdhost_flag_set(host, RRDHOST_FLAG_STREAM_LABELS_STOP);
+                rrdhost_flag_clear(host, RRDHOST_FLAG_STREAM_LABELS_UPDATE);
+            }
+            else {
+                stream_version = parse_stream_version_for_errors(http);
+            }
+        }
+    }
+    return stream_version;
+}
+
+static inline struct netdata_ssl *sender_ssl(struct sender_state *ss) {
+    return &ss->host->ssl;
+}
+
+static inline int sender_sockfd(struct sender_state *ss) {
+    return ss->host->rrdpush_sender_socket;
+}
+
+static inline RRDHOST *receiver_host(struct receiver_state *rs) {
+    return rs->host;
+}
+
+static inline struct netdata_ssl *receiver_ssl(struct receiver_state *rs) {
+    return &rs->ssl;
+}
+
+static inline int receiver_sockfd(struct receiver_state *rs) {
+    return rs->fd;
+}
+
+#define sender_error(sender, fmt, args...) do { \
+    size_t len = 1024; \
+    char newfmt[len]; \
+    memset(newfmt, 0, len); \
+    const char *connected_to = s->connected_to; \
+    snprintfz(newfmt, len, " [send to %%s]: %s", fmt); \
+    error(newfmt, connected_to, ##args); \
+} while(0);
+
+#define receiver_error(receiver, fmt, args...) do { \
+    size_t len = 1024; \
+    char newfmt[len]; \
+    memset(newfmt, 0, len); \
+    snprintfz(newfmt, len, " [recv from %%s (%%s:%%s)]: %s", fmt); \
+    error(newfmt, receiver_host(receiver), receiver->client_ip, receiver->client_port, ##args); \
+} while(0);
+
+static size_t dummy_tcp_data(struct netdata_ssl *ssl, int sockfd, time_t timeout,
+                           size_t processed, size_t expected, bool recv)
+{
+    if (processed > expected)
+        return ~0;
+
+    ssize_t remaining = expected - processed;
+    if (!remaining)
+        return 0;
+
+    char *buf = callocz(sizeof(char), remaining);
+    if (!buf)
```

Comment:
If `callocz` cannot allocate memory, it is expected that it will call `fatal`, so this test is not necessary. Considering that we are going to release this memory few lines below, do you think we can allocate a value bigger than the maximum allowed `stack`? Probably we could use a local variable here.